brainiak/searchlight/searchlight.py

```diff
@@ -0,0 +1,325 @@
+                only includes voxels within a 'rad' radius from the current
+                searchlight center
+                mask: A 3D numpy.ndarray object with type np.bool indicating
+                which includes voxels within a 'rad' radius from the current
+                searchlight center
+
+        rad: Odd positive integer indicating the radius of the searchlight
+        cube.
+
+        """
+
+        comm = MPI.COMM_WORLD
+        rank = comm.Get_rank()
+
+        # Request task (empty result)
+        comm.send((rank, None), dest=0)
+
+        # Infinite loop
+        while True:
+
+            # Receive task
+            (idx, d, m) = comm.recv(source=0)
+
+            # If task is empty then break loop
+            if(idx == -1):
+                break
+
+            result = fn(d, m)
+
+            # Send result
+            comm.send((rank, (idx, result)), dest=0)
+
+    def _dynamic_tasking(self, tasks, data, mask, fn, rad):
+        """ Distribute tasks dynamically to ranks 1-size
+
+        Parameters
+        ----------
+        tasks: searchlight centers, a list of tuples
+
+        data: A list of 4D numpy.ndarray objects which must have at least one
+        element. The computation is applied to this data.
+
+        mask: A 3D numpy.ndarray object with type np.bool indicating which
+        voxels are active in the searchlight operation.
+
+        fn: A user-provided function which is applied to the data in a
+        searchlight.
+            This user-specified function must accept 2 parameters:
+                data: A list of 4D numpy.ndarray objects which has the same
+                structure as the 'data' parameter to the searchlight, but
+                only includes voxels within a 'rad' radius from the current
+                searchlight center
+                mask: A 3D numpy.ndarray object with type np.bool indicating
+                which includes voxels within a 'rad' radius from the current
+                searchlight center
+
+        rad: Odd positive integer indicating the radius of the searchlight
+        cube.
+
+        """
+
+        comm = MPI.COMM_WORLD
+        rank = comm.Get_rank()
+
+        results = []
+
+        if rank == 0:
+            results = self._do_master(tasks, data, mask, fn, rad)
+
+        if rank != 0:
+            self._do_worker(tasks, data, mask, fn, rad)
+
+        return results
+
+    # Partition the mask and create tasks
+    def run(self, data, mask, fn, rad=1):
+        """ Run searchlight
+
+        Applies a function to each voxel present in the mask.
+
+        Parameters
+        ---------
+        data: A list of 4D numpy.ndarray objects which must have at least one
+        element. The computation is applied to this data.
+
+        mask: A 3D numpy.ndarray object with type np.bool indicating which
+                    voxels are active in the searchlight operation.
+
+        fn: A user-provided function which is applied to the data in a
```

Comment:
The user-provided function may require more parameters. For example, in Searchlight of FCMA toolbox, we need to pass in data labels, fold numbers for cross-validation, etc.